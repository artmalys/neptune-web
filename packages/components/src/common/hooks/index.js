export { default as useHasIntersected } from './useHasIntersected';
export { default as usePrevious } from './usePrevious';
