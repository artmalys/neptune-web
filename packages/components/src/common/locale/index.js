export const DEFAULT_LANG = 'en';
export const DEFAULT_LOCALE = 'en-GB';

const COUNTRY_ISO2_CODE_LENGTH = 2;

export const SUPPORTED_LANGUAGES = [
  DEFAULT_LANG,
  'ja',
  'de',
  'es',
  'fr',
  'hu',
  'id',
  'it',
  'pl',
  'pt',
  'ro',
  'ru',
  'tr',
  'zh',
];

export function adjustLocale(locale) {
  if (!locale || locale.trim().length === 0) {
    return null;
  }
  try {
    const { baseName } = new Intl.Locale(locale.replace('_', '-'));
    return baseName;
  } catch (e) {
    return null;
  }
}

export function mapLocaleToLanguage(locale) {
  const adjustedLocale = adjustLocale(locale);
  if (adjustedLocale === null) {
    return null;
  }
  try {
    const { language } = new Intl.Locale(adjustedLocale);
    if (SUPPORTED_LANGUAGES.includes(language)) {
      return language;
    }
    return null;
  } catch (e) {
    return null;
  }
}

export const localeService = {
  isValid: (locale) => {
    try {
      Intl.getCanonicalLocales(locale);
      return true;
    } catch {
      return false;
    }
  },
  getCountryFromLocale: (locale) => {
    if (localeService.isValid(locale)) {
      return locale.slice(0, COUNTRY_ISO2_CODE_LENGTH);
    }
    return null;
  },
};
