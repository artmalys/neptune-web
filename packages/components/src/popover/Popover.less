@import (reference) '@transferwise/neptune-css/src/less/variables/_typography.less';
.np-popover {
  &__content {
    font-size: @font-size-small;
  }
}
