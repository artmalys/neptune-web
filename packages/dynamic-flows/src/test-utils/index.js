import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount };
