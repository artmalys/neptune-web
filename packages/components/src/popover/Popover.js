import React, { useEffect, useState, useRef, useMemo } from 'react';
import Types from 'prop-types';
import classnames from 'classnames';
import uniqid from 'uniqid';

import { usePopper } from 'react-popper';
import { logActionRequiredIf } from '../utilities';

import { Position, Breakpoint } from '../common';
import BottomSheet from './BottomSheet';
import { useClientWidth, useIsClickOutside } from '../common/hooks';

import './Popover.css';

const Popover = ({ arrow, children, content, preferredPlacement, title }) => {
  logActionRequiredIf(
    `Popover has deprecated the ${preferredPlacement} value for the 'preferredPlacement' prop. Please use ${deprecatePlacements[preferredPlacement]} instead.`,
    deprecatePlacements[preferredPlacement],
  );

  const [arrowElement, setArrowElement] = useState(null);
  const [clientWidth] = typeof window !== 'undefined' ? useClientWidth({ ref: window }) : 0;
  const referenceElement = useRef(null);
  const outsideClickRef = useRef(null);

  const isMobile = clientWidth && clientWidth < Breakpoint.SMALL;
  const [popperElement, setPopperElement] = useState(null);

  const [isClickOutside] = useIsClickOutside({
    ref: [outsideClickRef, referenceElement],
  });

  const [open, setOpen] = useState(false);

  const memoizedId = useMemo(() => uniqid('np-popover-'), []);

  useEffect(() => {
    if (isClickOutside) {
      setOpen(false);
    }
  }, [isClickOutside]);

  const modifiers = [
    {
      name: 'computeStyles',
      options: {
        adaptive: false, // true by default
      },
    },
    { name: 'offset', options: { offset: [0, 12] } },
    { name: 'flip' },
  ];

  if (arrow) {
    modifiers.push({ name: 'arrow', options: { padding: 5, element: arrowElement } });
  }

  const placement = deprecatePlacements[preferredPlacement] || preferredPlacement;

  const { styles, attributes } = usePopper(referenceElement.current, popperElement, {
    placement,
    modifiers,
  });

  const handleClick = () => {
    setOpen(!open);
  };

  const popoverContent = (
    <div className="np-popover__content" id={memoizedId} aria-hidden={!open} role="tooltip">
      <span className="sr-only">Tooltip:</span>
      <div className="h5">{title}</div>
      {content}
    </div>
  );

  return (
    <>
      <span className="np-popover d-inline-block" ref={referenceElement}>
        {/* setting ref on child won't work for React components without forwardRef  */}
        {React.Children.map(children, (child) => {
          return React.cloneElement(child, {
            onClick: () => handleClick(),
            'aria-describedby': memoizedId,
          });
        })}
      </span>

      {isMobile ? (
        <BottomSheet open={open} ref={outsideClickRef}>
          {popoverContent}
        </BottomSheet>
      ) : (
        <div ref={outsideClickRef}>
          <div
            className={classnames('np-popover__dropdown', { 'np-popover__dropdown--open': open })}
            ref={setPopperElement}
            style={{ ...styles.popper }}
            {...attributes.popper}
          >
            {arrow && (
              <div
                className="np-popover__dropdown__arrow"
                ref={setArrowElement}
                style={styles.arrow}
              />
            )}
            {popoverContent}
          </div>
        </div>
      )}
    </>
  );
};

Popover.Placement = {
  TOP: Position.TOP,
  RIGHT: Position.RIGHT,
  BOTTOM: Position.BOTTOM,
  LEFT: Position.LEFT,
  LEFT_TOP: Position.LEFT_TOP,
  RIGHT_TOP: Position.RIGHT_TOP,
  BOTTOM_LEFT: Position.BOTTOM_LEFT,
  BOTTOM_RIGHT: Position.BOTTOM_RIGHT,
};

Popover.defaultProps = {
  arrow: true,
  preferredPlacement: Popover.Placement.TOP,
  title: undefined,
};

Popover.propTypes = {
  arrow: Types.bool,
  children: Types.node.isRequired,
  content: Types.node.isRequired,
  preferredPlacement: Types.oneOf([
    Popover.Placement.TOP,
    Popover.Placement.RIGHT,
    Popover.Placement.BOTTOM,
    /* @DEPRECATED Please use TOP instead. */
    Popover.Placement.LEFT_TOP,
    /* @DEPRECATED Please use TOP instead. */
    Popover.Placement.RIGHT_TOP,
    /* @DEPRECATED Please use BOTTOM instead. */
    Popover.Placement.BOTTOM_LEFT,
    /* @DEPRECATED Please use BOTTOM instead. */
    Popover.Placement.BOTTOM_RIGHT,
  ]),
  title: Types.string,
};

export const deprecatePlacements = {
  [Position.LEFT_TOP]: Popover.Placement.TOP,
  [Position.RIGHT_TOP]: Popover.Placement.TOP,
  [Position.BOTTOM_RIGHT]: Popover.Placement.BOTTOM,
  [Position.BOTTOM_LEFT]: Popover.Placement.BOTTOM,
};

export default Popover;
