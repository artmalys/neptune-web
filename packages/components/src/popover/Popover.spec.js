import React from 'react';
import { fireEvent, render, waitFor, screen } from '../test-utils';
import Popover from './Popover';
import { Breakpoint } from '../common';

jest.mock('./BottomSheet', () => {
  // eslint-disable-next-line react/prop-types
  return ({ open, children }) => {
    return open ? <div className="bottom-sheet">{children}</div> : null;
  };
});

describe('Popover', () => {
  const props = {
    arrow: true,
    content: 'content',
    intialOpen: false,
    placement: Popover.Placement.BOTTOM,
    title: 'title',
  };

  let container;
  describe('on desktop', () => {
    it('renders', async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props} intialOpen>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(container).toMatchSnapshot();
    });

    it('renders arrow', async () => {
      let rerender;
      await waitFor(() => {
        ({ container, rerender } = render(
          <Popover {...props} arrow={false}>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(getArrow()).not.toBeInTheDocument();

      await waitFor(() => {
        rerender(
          <Popover {...props} arrow>
            <button type="button">Open</button>
          </Popover>,
        );
      });

      expect(getArrow()).toBeInTheDocument();
    });

    it('opens/closes dropdown onClick', async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props}>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(getDropDown()).not.toBeInTheDocument();

      fireEvent.click(screen.getByText('Open'));

      expect(getDropDown()).toBeInTheDocument();

      fireEvent.click(screen.getByText('Open'));

      expect(getDropDown()).not.toBeInTheDocument();
    });

    it('close dropdown when clicking outside dropdown', async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props}>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(getDropDown()).not.toBeInTheDocument();

      fireEvent.click(screen.getByText('Open'));

      expect(getDropDown()).toBeInTheDocument();

      fireEvent.click(document);

      expect(getDropDown()).not.toBeInTheDocument();
    });

    it("doesn't close dropdown when clicking inside dropdown", async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props}>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(getDropDown()).not.toBeInTheDocument();

      fireEvent.click(screen.getByText('Open'));

      expect(getDropDown()).toBeInTheDocument();

      fireEvent.click(getDropDown());

      expect(getDropDown()).toBeInTheDocument();
    });
  });

  describe('on mobile', () => {
    beforeEach(() => {
      global.innerWidth = Breakpoint.SMALL - 1;
    });

    it('renders', async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props} initialOpen>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(container).toMatchSnapshot();
    });

    it('opens/closes bottomSheet onClick', async () => {
      await waitFor(() => {
        ({ container } = render(
          <Popover {...props}>
            <button type="button">Open</button>
          </Popover>,
        ));
      });

      expect(getBottomSheet()).not.toBeInTheDocument();

      fireEvent.click(screen.getByText('Open'));

      expect(getBottomSheet()).toBeInTheDocument();

      fireEvent.click(screen.getByText('Open'));

      expect(getBottomSheet()).not.toBeInTheDocument();
    });
  });

  const getDropDown = () => container.querySelector('.np-popover__dropdown--open');
  const getBottomSheet = () => container.querySelector('.bottom-sheet');
  const getArrow = () => container.querySelector('.np-popover__dropdown__arrow');
});
