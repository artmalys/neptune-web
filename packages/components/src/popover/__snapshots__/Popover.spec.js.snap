// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover on desktop renders 1`] = `
<div>
  <span
    class="d-inline-block"
  >
    <button
      type="button"
    >
      Open
    </button>
    <div
      class="np-popover"
    >
      <div>
        <div
          class="np-popover__dropdown"
          data-popper-escaped="true"
          data-popper-placement="bottom"
          data-popper-reference-hidden="true"
          style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 12px);"
        >
          <div
            class="np-popover__dropdown__arrow"
            style="position: absolute; left: 0px; transform: translate(5px, 0px);"
          />
          <div
            class="np-popover__content"
          >
            <div
              class="h5"
            >
              title
            </div>
            content
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;

exports[`Popover on mobile renders 1`] = `
<div>
  <span
    class="d-inline-block"
  >
    <button
      type="button"
    >
      Open
    </button>
    <div
      class="np-popover"
    >
      <div
        class="bottom-sheet"
      >
        <div
          class="np-popover__content"
        >
          <div
            class="h5"
          >
            title
          </div>
          content
        </div>
      </div>
    </div>
  </span>
</div>
`;
